q2cli/core/usage.py

```diff
@@ -0,0 +1,146 @@
+        return '\n'.join(self._recorder)
+
+    def get_example_data(self):
+        return {r: f() for r, f in self._init_data_refs.items()}
+
+    def extract(self, action_f, input_opts):
+        inputs = template_inputs(action_f, input_opts)
+        params = self.template_parameters(action_f, input_opts)
+        params += self.template_metadata()
+        return inputs, params
+
+    def _template_action(self, action_f, inputs, params, outputs):
+        cmd = to_cli_name(f"qiime {action_f.plugin_id} {action_f.id}")
+        cli_outputs = template_outputs(action_f, outputs)
+        t = " \\\n".join([cmd] + inputs + params + cli_outputs)
+        return t
+
+    def template_parameters(self, action_f, input_opts):
+        params = []
+        for i, spec in action_f.signature.parameters.items():
+            qtype = spec.qiime_type
+            val = str(input_opts[i]) if i in input_opts else ""
+            if spec.qiime_type is Bool:
+                pfx = f"--p-" if val == "True" else f"--p-no-"
+                p = f"{pfx}{to_cli_name(i)}"
+                params.append(f"{' ':>4}{p}")
+            elif not is_metadata_type(qtype) and val:
+                p = f"--p-{to_cli_name(i)}"
+                _val = f" {val}"
+                params.append(f"{' ':>4}{p + _val}")
+            elif is_metadata_column_type(qtype) and val not in self._col_refs:
+                self._col_refs[val] = val
+            elif is_metadata_type(qtype) and val not in self._merge_targets:
+                self._metadata_refs[val] = val
+        return params
+
+    def template_metadata(self):
+        params = []
+        for i in self._metadata_refs:
+            p = f"--m-metadata-file"
+            val = f"{i}.tsv"
+            params.append(f"{' ':>4}{p} {val}")
+        for i in self._col_refs:
+            if i not in self._metadata_refs:
+                p = f"--m-metadata-file"
+                val = f"{i}.tsv"
+                params.append(f"{' ':>4}{p} {val}")
+            col = self._col_refs[i]
+            p = f"--m-metadata-column"
+            params.append(f"{' ':>4}{p} '{col}'")
+        return params
+
+
+def template_inputs(action, input_opts):
+    inputs = []
+    for i in action.signature.inputs:
+        p = f"--i-{to_cli_name(i)}"
+        val = f"{input_opts[i]}.qza"
+        inputs.append(f"{' ':>4}{p} {val}")
+    return inputs
+
+
+def template_outputs(action, outputs):
+    cli_outputs = []
+    for i, spec in action.signature.outputs.items():
+        qtype = spec.qiime_type
+        ext = ".qzv" if is_visualization_type(qtype) else ".qza"
+        p = f"--o-{to_cli_name(i)}"
+        val = f"{to_snake_case(outputs[i])}{ext}"
+        cli_outputs.append(f"{' ':>4}{p} {val}")
+    return cli_outputs
+
+
+def examples(action):
+    all_examples = []
+    for i in action.examples:
+        use = CLIUsage()
+        action.examples[i](use)
+        example = use.render()
+        comment = f"# {i}".replace('_', ' ')
```

Comment:
It isn't super clear where one example begins and another ends, what if you do something like:

```bash
### example 1: concatenate ints simple ########################################

# This example demonstrates basic usage.
qiime dummy-plugin concatenate-ints \
    --i-ints1 ints_a.qza \
    --i-ints2 ints_b.qza \
    --i-ints3 ints_c.qza \
    --p-int1 4 \
    --p-int2 2 \
    --o-concatenated-ints ints_d.qza

###############################################################################

### example 2: concatenate ints complex #######################################

# This example demonstrates chained usage (pt 1).
qiime dummy-plugin concatenate-ints \
    --i-ints1 ints_a.qza \
    --i-ints2 ints_b.qza \
    --i-ints3 ints_c.qza \
    --p-int1 4 \
    --p-int2 2 \
    --o-concatenated-ints ints_d.qza
# This example demonstrates chained usage (pt 2).
qiime dummy-plugin concatenate-ints \
    --i-ints1 ints_d.qza \
    --i-ints2 ints_b.qza \
    --i-ints3 ints_c.qza \
    --p-int1 41 \
    --p-int2 0 \
    --o-concatenated-ints concatenated_ints.qza

###############################################################################

### example 3: comments only ##################################################

# comment 1
# comment 2
```